--- ../handmade/ray/ray.cpp	2017-12-07 19:18:40.771446179 +1100
+++ ray/ray.cpp	2017-12-10 15:20:19.818973867 +1100
@@ -426,7 +426,7 @@
     
     plane Planes[] =
     {
-        {{0, 0, 1}, {0}, 1},
+        {{0, 0, 1}, 0, 1},
     };
     
     sphere Spheres[] =
@@ -448,7 +448,7 @@
     
     image_u32 Image = AllocateImage(1280, 720);
     
-    u32 CoreCount = GetCPUCoreCount();
+    u32 CoreCount = 1; //GetCPUCoreCount();
     u32 TileWidth = Image.Width / CoreCount;
     u32 TileHeight = TileWidth;
     // TODO(casey): It seems like prescribing the tile to be 64x64 was the sweet spot
@@ -540,7 +540,8 @@
     }
     
     clock_t EndClock = clock();
-    clock_t TimeElapsed = EndClock - StartClock;
+    u32 clocksPerMS = CLOCKS_PER_SEC / 1000;
+    clock_t TimeElapsed = (EndClock - StartClock) / clocksPerMS;
     
     u64 UsedBounces = Queue.BouncesComputed;
     u64 TotalBounces = Queue.LoopsComputed;
@@ -562,6 +563,8 @@
 
 #if RAY_WIN32
 #include "win32_ray.cpp"
+#elif RAY_LINUX
+#include "linux_ray.cpp"
 #else
 #error "You need to define a platform!"
 #endif
